### ISSUE TITLE
Nix raises SIGBUS on `nix flake update`

### ISSUE THREAD

**Issue by zseri** [NixOS/nix#6212]
**Describe the bug**

```console
$ nix flake update
downloading 'https://api.github.com/repos/nix-community/home-manager/commits/HEAD'Bus error (core dumped)
```
(reproducably (e.g. not random) on this machine, it seems)

backtrace:
```
(gdb) bt
#0  0x00007f9affa0f137 in nix::trim[abi:cxx11](std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >) ()
   from /nix/store/qc9p0nkh3843zsx3ib4vl6z4azmz7f2y-nix-2.7.0pre20220225_df552ff/lib/libnixutil.so
#1  0x00007f9affc8f7d9 in nix::curlFileTransfer::TransferItem::headerCallback(void*, unsigned long, unsigned long) () from /nix/store/qc9p0nkh3843zsx3ib4vl6z4azmz7f2y-nix-2.7.0pre20220225_df552ff/lib/libnixstore.so
#2  0x00007f9afeeb3256 in Curl_client_write () from /nix/store/hh5ai1v2hhizkv5n4fr4hg72l1a5z1iw-curl-7.76.1/lib/libcurl.so.4
#3  0x00007f9afee981fe in Curl_http_readwrite_headers () from /nix/store/hh5ai1v2hhizkv5n4fr4hg72l1a5z1iw-curl-7.76.1/lib/libcurl.so.4
#4  0x00007f9afeec4cff in Curl_readwrite () from /nix/store/hh5ai1v2hhizkv5n4fr4hg72l1a5z1iw-curl-7.76.1/lib/libcurl.so.4
#5  0x00007f9afeeab374 in multi_runsingle () from /nix/store/hh5ai1v2hhizkv5n4fr4hg72l1a5z1iw-curl-7.76.1/lib/libcurl.so.4
#6  0x00007f9afeeac8ae in curl_multi_perform () from /nix/store/hh5ai1v2hhizkv5n4fr4hg72l1a5z1iw-curl-7.76.1/lib/libcurl.so.4
#7  0x00007f9affc8d370 in nix::curlFileTransfer::workerThreadMain() () from /nix/store/qc9p0nkh3843zsx3ib4vl6z4azmz7f2y-nix-2.7.0pre20220225_df552ff/lib/libnixstore.so
#8  0x00007f9affc8eb2c in nix::curlFileTransfer::workerThreadEntry() () from /nix/store/qc9p0nkh3843zsx3ib4vl6z4azmz7f2y-nix-2.7.0pre20220225_df552ff/lib/libnixstore.so
#9  0x00007f9aff77eca0 in ?? () from /nix/store/50msfhkz5wbyk8i78pjv3y9lxdrp7dlm-gcc-10.3.0-lib/lib/libstdc++.so.6
#10 0x00007f9aff520e9e in start_thread () from /nix/store/jsp3h3wpzc842j0rz61m5ly71ak6qgdn-glibc-2.32-54/lib/libpthread.so.0
#11 0x00007f9aff44b4af in clone () from /nix/store/jsp3h3wpzc842j0rz61m5ly71ak6qgdn-glibc-2.32-54/lib/libc.so.6
```

**Steps To Reproduce**
Unknown

**Expected behavior**
Nix shouldn't segfault while updating the lock file.

**`nix-env --version` output**
```console
$ nix-env --version
FOO nix-env
nix-env (Nix) 2.7.0pre20220225_df552ff
```

**Additional context**
```
experimental-features = nix-command ca-derivations flakes
```

Question: nixpkgs' `enableDebugging` doesn't appear to work on the nix flake provided by this repo, is there some workaround to still get debugging info? (or idk why the debug info isn't picked up by gdb... `set debug-info-directory` also doesn't work)

**Comment by zseri** [comment]
```
(No debugging symbols found in nix)
(gdb) set debug-info-directory /nix/store/6yzihyb8iw981m8fdc8cp3jjw6143pw8-nix-2.7.0pre20220225_df552ff-debug/lib/debug
No symbol table is loaded.  Use the "file" command.
```

**Comment by zseri** [comment]
```
(gdb) disas /m
Dump of assembler code for function _ZN3nix4trimB5cxx11ESt17basic_string_viewIcSt11char_traitsIcEES3_:
1264	{
   0x00007ffff74dc0c0 <+0>:	push   %r15
   0x00007ffff74dc0c2 <+2>:	push   %r14
   0x00007ffff74dc0c4 <+4>:	push   %r13
   0x00007ffff74dc0c6 <+6>:	push   %r12
   0x00007ffff74dc0c8 <+8>:	push   %rbp
   0x00007ffff74dc0c9 <+9>:	push   %rbx
   0x00007ffff74dc0ca <+10>:	sub    $0x28,%rsp
   0x00007ffff74dc0ce <+14>:	mov    %rdi,0x8(%rsp)
   0x00007ffff74dc0d3 <+19>:	mov    %fs:0x28,%rax
   0x00007ffff74dc0dc <+28>:	mov    %rax,0x18(%rsp)
   0x00007ffff74dc0e1 <+33>:	xor    %eax,%eax

1265	   auto i = s.find_first_not_of(whitespace);
1266	   if (i == std::string_view::npos) return "";
   0x00007ffff74dc133 <+115>:	lea    -0x1(%r13),%rbx
=> 0x00007ffff74dc137 <+119>:	movsbl 0x0(%rbp,%rbx,1),%esi
   0x00007ffff74dc13c <+124>:	mov    %r14,%rdx
   0x00007ffff74dc13f <+127>:	mov    %r15,%rdi
   0x00007ffff74dc142 <+130>:	call   0x7ffff7448070 <memchr@plt>
   0x00007ffff74dc147 <+135>:	test   %rax,%rax
   0x00007ffff74dc14a <+138>:	je     0x7ffff74dc1f7 <_ZN3nix4trimB5cxx11ESt17basic_string_viewIcSt11char_traitsIcEES3_+311>
   0x00007ffff74dc150 <+144>:	sub    $0x1,%rbx
   0x00007ffff74dc154 <+148>:	jae    0x7ffff74dc137 <_ZN3nix4trimB5cxx11ESt17basic_string_viewIcSt11char_traitsIcEES3_+119>
   0x00007ffff74dc156 <+150>:	cs nopw 0x0(%rax,%rax,1)
   0x00007ffff74dc160 <+160>:	cmp    %r13,%r12
   0x00007ffff74dc163 <+163>:	ja     0x7ffff74dc26b <_ZN3nix4trimB5cxx11ESt17basic_string_viewIcSt11char_traitsIcEES3_+427>
   0x00007ffff74dc169 <+169>:	mov    0x8(%rsp),%rax
   0x00007ffff74dc16e <+174>:	sub    %r12,%r13
   0x00007ffff74dc171 <+177>:	cmp    %rbx,%r13
   0x00007ffff74dc174 <+180>:	cmova  %rbx,%r13
   0x00007ffff74dc178 <+184>:	lea    0x10(%rax),%rdi
   0x00007ffff74dc17c <+188>:	add    %r12,%rbp
--Type <RET> for more, q to quit, c to continue without paging--c
   0x00007ffff74dc17f <+191>:	mov    %rdi,(%rax)
   0x00007ffff74dc182 <+194>:	mov    %rbp,%rax
   0x00007ffff74dc185 <+197>:	add    %r13,%rax
   0x00007ffff74dc188 <+200>:	je     0x7ffff74dc193 <_ZN3nix4trimB5cxx11ESt17basic_string_viewIcSt11char_traitsIcEES3_+211>
   0x00007ffff74dc18a <+202>:	test   %rbp,%rbp
   0x00007ffff74dc18d <+205>:	je     0x7ffff74dc286 <_ZN3nix4trimB5cxx11ESt17basic_string_viewIcSt11char_traitsIcEES3_+454>
   0x00007ffff74dc193 <+211>:	mov    %r13,0x10(%rsp)
   0x00007ffff74dc198 <+216>:	cmp    $0xf,%r13
   0x00007ffff74dc19c <+220>:	ja     0x7ffff74dc210 <_ZN3nix4trimB5cxx11ESt17basic_string_viewIcSt11char_traitsIcEES3_+336>
   0x00007ffff74dc19e <+222>:	cmp    $0x1,%r13
   0x00007ffff74dc1a2 <+226>:	jne    0x7ffff74dc208 <_ZN3nix4trimB5cxx11ESt17basic_string_viewIcSt11char_traitsIcEES3_+328>
   0x00007ffff74dc1a4 <+228>:	movzbl 0x0(%rbp),%eax
   0x00007ffff74dc1a8 <+232>:	mov    0x8(%rsp),%rcx
   0x00007ffff74dc1ad <+237>:	mov    %al,0x10(%rcx)
   0x00007ffff74dc1b0 <+240>:	mov    0x8(%rsp),%rax
   0x00007ffff74dc1b5 <+245>:	mov    %r13,0x8(%rax)
   0x00007ffff74dc1b9 <+249>:	movb   $0x0,(%rdi,%r13,1)
   0x00007ffff74dc250 <+400>:	mov    0x8(%rsp),%rdi
   0x00007ffff74dc255 <+405>:	lea    0x1cd59(%rip),%rsi        # 0x7ffff74f8fb5
   0x00007ffff74dc25c <+412>:	call   0x7ffff74db3b0 <_ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEC4IS3_EEPKcRKS3_>
   0x00007ffff74dc261 <+417>:	jmp    0x7ffff74dc1be <_ZN3nix4trimB5cxx11ESt17basic_string_viewIcSt11char_traitsIcEES3_+254>

1267	   auto j = s.find_last_not_of(whitespace);
1268	   return std::string(s, i, j == std::string::npos ? j : j - i + 1);
   0x00007ffff74dc1f7 <+311>:	sub    %r12,%rbx
   0x00007ffff74dc1fa <+314>:	add    $0x1,%rbx
   0x00007ffff74dc1fe <+318>:	jmp    0x7ffff74dc160 <_ZN3nix4trimB5cxx11ESt17basic_string_viewIcSt11char_traitsIcEES3_+160>
   0x00007ffff74dc203 <+323>:	nopl   0x0(%rax,%rax,1)

1269	}
   0x00007ffff74dc1be <+254>:	mov    0x18(%rsp),%rax
   0x00007ffff74dc1c3 <+259>:	sub    %fs:0x28,%rax
   0x00007ffff74dc1cc <+268>:	jne    0x7ffff74dc266 <_ZN3nix4trimB5cxx11ESt17basic_string_viewIcSt11char_traitsIcEES3_+422>
   0x00007ffff74dc1d2 <+274>:	mov    0x8(%rsp),%rax
   0x00007ffff74dc1d7 <+279>:	add    $0x28,%rsp
   0x00007ffff74dc1db <+283>:	pop    %rbx
   0x00007ffff74dc1dc <+284>:	pop    %rbp
   0x00007ffff74dc1dd <+285>:	pop    %r12
   0x00007ffff74dc1df <+287>:	pop    %r13
   0x00007ffff74dc1e1 <+289>:	pop    %r14
   0x00007ffff74dc1e3 <+291>:	pop    %r15
   0x00007ffff74dc1e5 <+293>:	ret    
   0x00007ffff74dc1e6 <+294>:	cs nopw 0x0(%rax,%rax,1)
   0x00007ffff74dc266 <+422>:	call   0x7ffff7445470 <__stack_chk_fail@plt>
```

**Comment by zseri** [comment]
hm, in my naive view this appears to be a miscompilation, idk.

**Comment by puckipedia** [comment]
Not a miscompilation, and can be reliably repeated here too:

```
marisa ~> /nix/store/qc9p0nkh3843zsx3ib4vl6z4azmz7f2y-nix-2.7.0pre20220225_df552ff/bin/nix repl
Welcome to Nix 2.7.0pre20220225_df552ff. Type :? for help.

nix-repl> builtins.fetchurl "https://http2.akamai.com/demo"
downloading 'https://http2.akamai.com/demo'fish: Job 1, '/nix/store/qc9p0nkh3843zsx3ib4v…' terminated by signal SIGBUS (Misaligned address error)
```

Probably a misparsing of the HTTP/2 status line :)

**Comment by edolstra** [comment]
Try upgrading, this should be fixed in 609779086301a600f4282a629d586bba6b6a485c.

**Comment by zseri** [comment]
I upgraded to 92b8d4d8861b908a7ec500526a84155c597d6d2b and the bug appears to be gone, ok.
